src/charm.py

```diff
@@ -84,100 +79,101 @@ def _on_config_changed(self, event):
         self._configure(event)
 
     def _configure(self, event):
-        restart = False
-
         if not self._container.can_connect():
             self.unit.status = WaitingStatus("Waiting for Pebble ready")
             return
 
         try:
-            restart = self._set_alerts(self._container)
-        except PebbleError as e:
-            logger.error("Failed to push updated alert files: %s", e)
-            self.unit.status = BlockedStatus("Failed to push updated alert files; see debug logs")
+            self._set_alerts()
+            self._set_mimir_config()
+            self._set_pebble_layer()
+        except BlockedStatusError as e:
+            self.unit.status = BlockedStatus(e.message)
             return
 
+        if self._restart:
+            self._container.restart(self._name)
+            logger.info("Mimir (re)started")
+
+        self.remote_write_provider.update_endpoint()
+        self.unit.status = ActiveStatus()
+
+    def _set_mimir_version(self) -> bool:
+        version = self._mimir_version
+
+        if version is None:
+            logger.debug("Cannot set workload version at this time: could not get Mimir version.")
+            return False
+
+        self.unit.set_workload_version(version)
+        return True
+
+    def _set_pebble_layer(self):
         current_layer = self._container.get_plan().to_dict()
         new_layer = self._pebble_layer
 
         if "services" not in current_layer or DeepDiff(
             current_layer["services"], new_layer["services"], ignore_order=True
         ):
-            restart = True
+            self._container.add_layer(self._name, new_layer, combine=True)
+            self._restart = True
```

Comment:
[Iiuc](https://github.com/canonical/operator/issues/863), in the **rare** case that `self._container.restart(self._name)` raises (e.g. ChangeError) after a `can_connect` guard, the workload will not restart even after a `resolve`.

Unless `get_plan` gives you the stuff that's currently running, i.e. `add_layer` does not affect the output of `get_plan`. The docstring could be interpreted either way:
```python
"""Get the current effective pebble configuration."""
```

Reply:
AFAIU the `.restart` method will eventually raise an [`pebble.APIError`](https://github.com/canonical/operator/blob/main/ops/model.py#L1459)...